### ISSUE TITLE
Provide a REPL plugin that lets you drive a running fastify server from the command line.

### ISSUE THREAD

**Issue by rjrodger** [fastify/fastify#4961]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

Provide a plugin that exposes a REPL (the Node.js REPL) on a port on a running fastify instance. Users can then debug the server by running JS statements, injecting requests, and directly inspecting state. This is very useful for local dev work.

A proof-of-concept with some sample convenience commands is implemented in this gist: https://gist.github.com/rjrodger/2de69753bb670316b4395c10bf4cd6b7

See the example section for a sample interaction.

I am happy to continue work on this code and collaborate to turn into real plugin.

### Motivation

In the https://github.com/senecajs/seneca project have been using a REPL plugin for many years: https://github.com/senecajs/seneca-repl

This has given us a pretty awesome developer experience in local environments - it provides a high utility interface for initial implementations, debugging, and general messing around with a running server (throw in hot reloading plugins for extra fun). 

Recently we have also taken this into deployed environments (with suitable protections), and even into Serverless functions (the PoC code does not deal with this case--you need to abstract the streams). This is super helpful when stuff breaks on staging, but worked locally.

(And @mcollina calling my bluff! https://twitter.com/matteocollina/status/1686789782762909706)

### Example

This is an example interaction session against the following server (see gist for REPL code):
```
Fastify
  .get('/', function handler (request, reply) {
    reply.send({ hello: 'world' })
  })
  .get('/foo', function handler (request, reply) {
    reply.send('FOO')
  })
  .post('/bar', function handler (request, reply) {
    reply.send({...request.body, when:Date.now()})
  })
```

On the command line, using telnet (a REPL client will also be needed, because `node:readline` is a must have in the real world).

```sh
$ telnet 127.0.0.1 31313
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.

fastify-4.21.0> 1+1                                        # Just another Node REPL!
2

fastify-4.21.0> fastify.version                            # Look, there's fastify!
'4.21.0'

fastify-4.21.0> get /                                      # Some quick syntax for injections
{ statusCode: 200, body: '{"hello":"world"}' }

fastify-4.21.0> get /foo                                   # Grammar might be: HTTP-METHOD URL [BODY] [OPTIONS]
{ statusCode: 200, body: 'FOO' }

fastify-4.21.0> post /bar {x:1}                            # Needs to be extended for headers etc.
{ statusCode: 200, body: '{"x":1,"when":1691526437680}' }

fastify-4.21.0> inject({method:'GET',url:'/foo'})          # Inject as a function
{
  raw: { ... },
  req: Request { ... },
  headers: {
    'content-type': 'text/plain; charset=utf-8',
    'content-length': '3',
    date: 'Tue, 08 Aug 2023 20:27:36 GMT',
    connection: 'keep-alive'
  },
  statusCode: 200,
  statusMessage: 'OK',
  trailers: {},
  cookies: [Getter],
  rawPayload: <Buffer 46 4f 4f>,
  payload: 'FOO',
  body: 'FOO',
  json: [Function: parseJsonPayload]
}
```

The REPL commands would need to be thought out a little more, of course. And the result object.

**Comment by rjrodger** [comment]
Oh yes. Absolutely. Bear in mind the primary use case is local dev work however.

An actual plugin would need some detailed work to lock things down. "With great power comes great responsibility" etc.

**Comment by jsumners** [comment]
How is this better than using existing tooling like https://curl.se or https://github.com/Orange-OpenSource/hurl?

**Comment by mcollina** [comment]
I would personally leave out the jsonic + custom handling for inject and just rely on providing access to the Fastify server, so that a user could call `server.inject()` themselves. Wdyt?

It would be an amazing plugin to add!

**Comment by rjrodger** [comment]
@mcollina yeah the custom command language is not the essential part - just a little demo of what could be possible - I would say though we have found it to be pretty great for DX in seneca projects 

@jsumners the real utility here is not so much doing injections, but introspecting and manipulating the business logic code implementing the requests. If you have a bunch of business logic plugins (~ microservices) that "do stuff", you can futz about with their internal data directly, call business methods etc, rather than just running a request from the top

**Comment by jsumners** [comment]
> but introspecting and manipulating the business logic code implementing the requests

How would you do this?

**Comment by rjrodger** [comment]
Depends entirely on how you've put together your business logic. Here's a hypothetical scenario: let's say `users` is some sort of representation of my registered users, and I'm working on access rules of some kind. A debug interaction might be:

```
> users.find('alice')
{id:'123',handle:'alice',role:'guest'}
> alice = _  # Node REPL feature: _ is the value of the last result
> users.login(alice) # let's say we have login method
> TOKEN = _.token # result =~ { ..., token: '...', ... }
> get /api/admin/metrics {Authorization: 'Bearer [TOKEN]'} # string substitution, say
{ statusCode: 401, body: 'Not Allowed' }
> alice.role = 'admin'
> alice.save() # guess we're using an ORM!
> get /api/admin/metrics {Authorization: 'Bearer [TOKEN]'}
{ statusCode: 200, body: { metrics: ... } }
```

The REPL means you can do this ad hoc and experiment. Perhaps convert it to a test later.

**Comment by Uzlopak** [comment]
Yeah, I find the idea also intriguing.

**Comment by rjrodger** [comment]
So long as you have the `fastify` instance and whatever business objects make sense, you are indeed good to go with plain JS.

As noted above, the real work here is on the security side - making sure the REPL is default "off" etc.

**Comment by jsumners** [comment]
> Depends entirely on how you've put together your business logic.

My question was really more along the lines of:

You said "manipulating the business logic code". How would you do that in this REPL? Are you replacing functions? What does that look like?

**Comment by Uzlopak** [comment]
The issue was jsumners is pointing to is, that after we let the instance listen it wont be modified that useful anymoree.

**Comment by mcollina** [comment]
This kind of REPL was popularized by Ruby On Rails.
This would be an awesome plugin to have!

**Comment by andersonjoseph** [comment]
+1 this would be an interesting and useful option to have. 

Interested in collaborate in what I can to bring this as a real plugin

**Comment by mcollina** [comment]
Should I set up a repo on this org?

**Comment by rjrodger** [comment]
@mcollina yes - happy to contribute - although I am not up to speed with fastify community conventions - will need some guidance

**Comment by cesconix** [comment]
I'm really excited about the idea of having a REPL for Fastify. I would be very interested in contributing.
